### ISSUE TITLE
object:set_properties() generates excess network packets

### ISSUE THREAD

**Issue by Warr1024** [minetest/minetest#9934]
##### Minetest version
```
Minetest 5.3.0-dev-e79bc40c0 (Linux)
Using Irrlicht 1.8.4
BUILD_TYPE=Release
RUN_IN_PLACE=1
USE_GETTEXT=0
USE_SOUND=1
USE_CURL=1
USE_FREETYPE=1
USE_LUAJIT=0
STATIC_SHAREDIR="."
```

##### OS / Hardware
Operating system: `Linux 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1+deb10u1 (2020-04-27) x86_64 GNU/Linux`
CPU: `Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz`
GPU model: `Mesa DRI Intel(R) Ivybridge Mobile  (0x166)`
OpenGL version: `4.2 (Core Profile) Mesa 18.3.6`

##### Summary

Calls to object:set_properties({...}) always unconditionally send a TOCLIENT_ACTIVE_OBJECT_MESSAGES message on that tick, even if the properties being set are the same as the old properties.  These packets are large, and for mods/games that involve a lot of entities, it can have a major impact on very bandwidth-constrained clients.  This forces mod authors to maintain complex and fragile workarounds to detect when values being set already match current values on the object (example: https://gitlab.com/sztest/nodecore/-/blob/master/mods/nc_api_ents/util.lua#L22).

The engine already properly maintains the properties of objects, and properly transfers existing state as entities enter/leave a client's visible space, so there is no need for mod authors to be able to _force_ a TOCLIENT_ACTIVE_OBJECT_MESSAGES packet send explicitly.  The engine should just automatically send this packet **only** when there are actual property changes that the client doesn't already know about.

##### Steps to reproduce

Install this demo mod:

```lua
local modname = minetest.get_current_modname()

minetest.register_entity(modname .. ":test", {
	initial_properties = {
		nametag = "quiet",
		static_save = false
	},
	on_step = function(self)
		if self.noise then self.object:set_properties({nametag = "noisy"}) end
	end,
	on_punch = function(self)
		self.noise = not self.noise
		if not self.noise then self.object:set_properties({nametag = "quiet"}) end
	end
})

minetest.register_on_joinplayer(function(player)
	local pos = vector.add(player:get_pos(), player:get_look_dir())
	pos.y = pos.y + player:get_properties().eye_height
	minetest.add_entity(pos, modname .. ":test")
end)
```

Enable the mod and start up the game.

Open the F5 profiler graphs and allow the area around the player to quiesce, finish loading mapblocks in the area, etc.  You should see the `client_received_packets` graph mostly empty, and there should be gaps with no packets received.

Punch the quiet/noisy entity in front of you to toggle its state.

##### Observed

When the entity is in "noisy" state, you will see an extra packet received each tick, compared to the quiet state.

##### Expected

Since the properties being set by the entity while in the noisy state match the values already set, there should be no additional noise regardless of the entity's state.  The `client_received_packets` graph should look the same regardless of whether the entity is in the "quiet" or "noisy" state.

**Comment by sfan5** [comment]
I am inclined to say this is not a bug.
E.g. the same applies to the following calls for players:
* `set_fov`
* `set_inventory_formspec`
* `set_physics_override`
* `hud_change`
* `hud_set_hotbar_`*
* `set_sky`, `set_sun`, `set_moon`, `set_stars`
* `set_clouds`
* `set_eye_offset`

The only caching LuaEntitySAO does is for position, velocity, acceleration and rotation.

**Comment by Warr1024** [comment]
> E.g. the same applies to the following calls for players

Sounds more like "this bug is bigger than you think" than "not a bug."

I mightn't have reported this except that it has caused a good measure of pain across multiple incidents, the cause was not intuitive, it's impractical to work around this in mod code, and mod-level patching is haphazard.

Also, I don't see how it's possible for the engine to actually _not_ cache these properties, since they're sent properly to a client who enters an area where an existing entity is loaded (e.g. by another player being present), without having to call set_properties() from the mod code.  Wouldn't this require the properties to be stored by the engine somewhere?

> Wouldn't it be possible to structure your code so that it sets properties event-based (when a change actually happens) instead of the current approach?

The engine cannot make assumptions about a modder's current approach.  As it is, I'm already managing things based on "events", but I have to poll as a fallback to handle situations for which events don't exist, like when a downstream mod makes a change to inventory/metadata in the node in which the entity exists.  There are too many situations that may cause an entity to need to update to try to trap them all, and it's impractical (at least) to try to enforce downstream mods not using a subset of the engine API that has known issues or could cause event triggers to be missed.

**Comment by Warr1024** [comment]
It turns out that I was able to modify my workaround to monkey-patch the API, instead of requiring modders to use an alternate API, so this could conceivably be done in the engine's lua builtins.

https://gitlab.com/sztest/nodecore/-/blob/e7f6afd789747d757047cb9b49f5b527c0cbf688/mods/nc_api_ents/util.lua

**Comment by SmallJoker** [comment]
Related: #6572

Relevant serialization code:
https://github.com/minetest/minetest/blob/5311a275235d2132443643199330953c877b794a/src/object_properties.cpp#L75-L77

For a player that's around 180 bytes, dropped items it's 250 to 370 bytes.
From there, only strings (such as textures and -modifiers) result in more transferred data. With 20 people in range that's `370 bytes * 20 / 0.09s ≃ 80.3 KiB/s`.

**Comment by paramat** [comment]
> Calls to object:set_properties({...}) always unconditionally send a TOCLIENT_ACTIVE_OBJECT_MESSAGES message on that tick, even if the properties being set are the same as the old properties.

I have mentioned something similar before in another issue or PR, but for the player object, the fact that all player properties are reset, including mesh reloading and animation reset, even if only 1 property has actually changed.
It was necessary to use this code in MTG to avoid visually unpleasant mesh resets every few seconds:
```
	-- Only set property if necessary to avoid player mesh reload
	if player:get_properties().zoom_fov ~= new_zoom_fov then
		player:set_properties({zoom_fov = new_zoom_fov})
	end
```
So i support doing something to improve how this works.
I consider it more a request for improving inconvenient code than a 'bug'.

**Comment by paramat** [comment]
See #6572

**Comment by sfan5** [comment]
> Also, I don't see how it's possible for the engine to actually _not_ cache these properties, since they're sent properly to a client who enters an area where an existing entity is loaded (e.g. by another player being present), without having to call set_properties() from the mod code. Wouldn't this require the properties to be stored by the engine somewhere?

You're correct that the engine stores the properties internally. In fact implementing what you suggested should be pretty simple, because it already stores them.

> (2) not all of those have a getter that allows the workaround to be centralized instead of maintaining caches everywhere

The ones without getters are probably the ones where the engine currently *doesn't* store the value internally, e.g. this is the case for `hud_change`.
